### PR TITLE
Promote sources API to v1

### PR DESCRIPTION
**Pull request by danyinggu**
Helps with #1409, #1410, #1411, #1412, #1413.

## Proposed Changes

- 🎁  Promote defaults, lifecycle types and validataion APIs of CloudPubsub, CloudBuild, CloudScheduler and CloudAuditsLogSource to v1.
- No logic change from v1beta1. Currently no code is using this API.
-

**Release Note**

```release-note
NONE
```

**Docs**

### CODE REVIEW

File: pkg/apis/events/v1/cloudbuildsource_defaults.go [hunk 1]
```diff
@@ -0,0 +1,37 @@
+/*
+Copyright 2020 Google LLC.
```
Comment (nachocano) [inline pkg/apis/events/v1/cloudbuildsource_defaults.go:2]: agreed offline to remove all CloudBuildSource files

File: pkg/apis/events/v1/cloudpubsubsource_defaults.go [hunk 2]
```diff
@@ -0,0 +1,52 @@
...
+	"github.com/google/knative-gcp/pkg/apis/duck"
+
+	"knative.dev/pkg/apis"
+	"knative.dev/pkg/ptr"
+)
+
+const (
+	defaultRetentionDuration = 7 * 24 * time.Hour
+	defaultAckDeadline       = 30 * time.Second
+)
```
Comment (nachocano) [inline pkg/apis/events/v1/cloudpubsubsource_defaults.go:32]: IIRC we had this in pullsubscription and you moved them to register.go
If that is the case, can you import and use those?

File: pkg/apis/events/v1/cloudpubsubsource_types.go [hunk 3]
```diff
@@ -0,0 +1,176 @@
...
+	_ resourcesemantics.GenericCRD = (*CloudPubSubSource)(nil)
+	_ kngcpduck.Identifiable       = (*CloudPubSubSource)(nil)
+	_ kngcpduck.PubSubable         = (*CloudPubSubSource)(nil)
+	_ duckv1.KRShaped              = (*CloudPubSubSource)(nil)
+)
+
+// CloudPubSubSourceSpec defines the desired state of the CloudPubSubSource.
+type CloudPubSubSourceSpec struct {
+	// This brings in the PubSub based Source Specs. Includes:
+	// Sink, CloudEventOverrides, Secret, PubSubSecret, and Project
```
Comment (nachocano) [inline pkg/apis/events/v1/cloudpubsubsource_types.go:61]: nit: PubSubSecret is not there anymore.

File: pkg/apis/events/v1/cloudpubsubsource_validation.go [hunk 4]
```diff
@@ -0,0 +1,101 @@
...
+	"knative.dev/pkg/apis"
+)
+
+const (
+	minRetentionDuration = 10 * time.Second   // 10 seconds.
+	maxRetentionDuration = 7 * 24 * time.Hour // 7 days.
+
+	minAckDeadline = 0 * time.Second  // 0 seconds.
+	maxAckDeadline = 10 * time.Minute // 10 minutes.
+)
```
Comment (nachocano) [inline pkg/apis/events/v1/cloudpubsubsource_validation.go:38]: same here

File: pkg/apis/events/v1/cloudbuildsource_defaults.go (hunk 1 above)
Comment (danyinggu) [reply pkg/apis/events/v1/cloudbuildsource_defaults.go]: removed.

File: pkg/apis/events/v1/cloudpubsubsource_validation.go [hunk 5]
```diff
@@ -0,0 +1,101 @@
...
+
+	if current.AckDeadline != nil {
+		// If set, AckDeadline needs to parse to a valid duration.
+		ad, err := time.ParseDuration(*current.AckDeadline)
+		if err != nil {
+			errs = errs.Also(apis.ErrInvalidValue(*current.AckDeadline, "ackDeadline"))
+		} else if ad < minAckDeadline || ad > maxAckDeadline {
+			errs = errs.Also(apis.ErrOutOfBoundsValue(*current.AckDeadline, minAckDeadline.String(), maxAckDeadline.String(), "ackDeadline"))
+		}
+	}
```
Comment (nachocano) [inline pkg/apis/events/v1/cloudpubsubsource_validation.go:76]: if would be nice if we can deduplicate this code. It's both in PullSubscription and here. 
seems like a good candidate to have a validations.go helper inside apis/intevents... and we can import it from here and use it... 
Same thing will do the PullSubscription

File: pkg/apis/events/v1/cloudauditlogssource_types.go [hunk 6]
```diff
@@ -0,0 +1,138 @@
...
+	_ apis.Validatable             = (*CloudAuditLogsSource)(nil)
+	_ kmeta.OwnerRefable           = (*CloudAuditLogsSource)(nil)
+	_ resourcesemantics.GenericCRD = (*CloudAuditLogsSource)(nil)
+	_ kngcpduck.Identifiable       = (*CloudAuditLogsSource)(nil)
+	_ kngcpduck.PubSubable         = (*CloudAuditLogsSource)(nil)
+	_ duckv1.KRShaped              = (*CloudAuditLogsSource)(nil)
+)
+
+const (
+	SinkReady apis.ConditionType = "SinkReady"
```
Comment (nachocano) [inline pkg/apis/events/v1/cloudauditlogssource_types.go:56]: let's add a comment:

// SinkReady has status True when CloudAuditLogsSource sink is ready.

File: pkg/apis/events/v1/cloudpubsubsource_defaults.go (hunk 2 above)
Comment (danyinggu) [reply pkg/apis/events/v1/cloudpubsubsource_defaults.go:32]: The pullsubscription PR #1441 is still pending. Can we do the refactor in a following PR in the future? As the API is blocking the reconciler stuff.

File: pkg/apis/events/v1/cloudpubsubsource_validation.go (hunk 5 above)
Comment (danyinggu) [reply pkg/apis/events/v1/cloudpubsubsource_validation.go:76]: Same thing as the pullsubscription PR is still pending

File: pkg/apis/events/v1/cloudauditlogssource_validation.go [hunk 7]
```diff
@@ -0,0 +1,77 @@
...
+
+	return errs
+}
+
+func (current *CloudAuditLogsSource) CheckImmutableFields(ctx context.Context, original *CloudAuditLogsSource) *apis.FieldError {
+	if original == nil {
+		return nil
+	}
+
+	// Modification of Topic, Secret, ServiceAccount, Project, ServiceName, MethodName, and ResourceName are not allowed. Everything else is mutable.
```
Comment (capri-xiyue) [inline pkg/apis/events/v1/cloudauditlogssource_validation.go:66]: Should it be `ServiceAccountName`?

File: pkg/apis/events/v1/cloudauditlogssource_validation_test.go [hunk 8]
```diff
@@ -0,0 +1,308 @@
...
+						Sink: auditLogsSourceSpec.PubSubSpec.Sink,
+					},
+				},
+				MethodName:   auditLogsSourceSpec.MethodName,
+				ResourceName: auditLogsSourceSpec.ResourceName,
+				ServiceName:  auditLogsSourceSpec.ServiceName,
+			},
+			allowed: false,
+		},
+		"ServiceAccount changed": {
```
Comment (capri-xiyue) [inline pkg/apis/events/v1/cloudauditlogssource_validation_test.go:227]: nit: "ServiceAccountName changed"

File: pkg/apis/events/v1/cloudpubsubsource_defaults.go (hunk 2 above)
Comment (capri-xiyue) [reply pkg/apis/events/v1/cloudpubsubsource_defaults.go:32]: https://github.com/google/knative-gcp/pull/1441 already got merged.

Comment (danyinggu) [reply pkg/apis/events/v1/cloudpubsubsource_defaults.go:32]: I will fix this with #1441's change tomorrow in this PR.

File: pkg/apis/events/v1/cloudpubsubsource_validation.go (hunk 5 above)
Comment (danyinggu) [reply pkg/apis/events/v1/cloudpubsubsource_validation.go:76]: Same as above. This will also be fixed in this PR.

**Comment by danyinggu** [comment]
Github seems to not pick up my new commits, I have to create a new PR #1462 with all fixes of the addressing comments.

**Pull request closed without merging**